Add settings to enable 24-hour time format and customizable date format

GimleLarpes opened the pull request:
<!-- Please read https://github.com/ajbura/cinny/blob/dev/CONTRIBUTING.md before submitting your pull request -->

### Description
<!-- Please include a summary of the change. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
Based on #1739  and #2344
Options for configuring date and time formats have long been requested (see #60 and duplicates). This PR aims to add this functionality.

#### This PR adds:
- A "Date & Time" subsection in General settings
- A toggle that allows the user to switch between 12-hour and 24-hour time formats
- A dropdown/text input field to let the user select a date format

As requested in #60.

In addition to the current `D MMM YYYY` format the following options are added: `DD/MM/YYYY`, `MM/DD/YYYY`, `YYYY/MM/DD`, `Custom` - where the user can enter a 16-character dayjs format string.


#### Example image:
![image](https://github.com/user-attachments/assets/e3a7346d-3b8b-4427-85ad-a304366b898c)
<br />

Fixes #60 

#### Type of change

- [o] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) (new function params)
- [o] This change requires a documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

ajbura commented on src/app/features/settings/general/General.tsx:
```diff
@@ -341,6 +345,224 @@ function Appearance() {
 
+type CustomDateFormatProps = {
+  dateFormatString: string;
+  setDateFormatString: (format: string) => void;
```
can we rename these to `value` and `onChange`? as this component is all about custom date format, naming props the same makes it little hard to read the internals of component.

ajbura commented on src/app/features/settings/general/General.tsx:
```diff
@@ -341,6 +345,224 @@ function Appearance() {
+
+type PresetDateFormatProps = {
+  dateFormatPreset: string;
+  handlePresetChange: (format: string) => void;
```
these can also be renamed to `value` and `onChange`

ajbura commented on src/app/features/settings/general/General.tsx:
```diff
@@ -341,6 +345,224 @@ function Appearance() {
+                    radii="300"
+                    onClick={() => handleSelect(item.format)}
+                  >
+                    <Text size="T300">{item.name}</Text>
```
i was wondering will it be a good idea to render the format here instead of format literals? and we do not need it render it in description unless "Custom" option is selected, and the change will be more clear to user.

ajbura commented on src/app/features/settings/general/General.tsx:
```diff
@@ -341,6 +345,224 @@ function Appearance() {
+
+function SelectDateFormat() {
+  const [dateFormatString, setDateFormatString] = useSetting(settingsAtom, 'dateFormatString');
+  const [dateFormatPreset, setDateFormatPreset] = useState(dateFormatString);
```
we can rename this to `selectedDateFormat` and `setSelectedDateFormat`

ajbura reviewed (changes requested):
Overall looks really solid!

kfiven commented on src/app/features/settings/general/General.tsx:
```diff
@@ -341,6 +345,224 @@ function Appearance() {
+        <Text size="T200">
+          View formatting syntax on{' '}
+          <a
+            href="https://day.js.org/docs/en/display/format"
```
IMHO we should remove the custom option as this makes dayjs a hard dependency and links to 3rd party sites in interface should be avoided. Also it’s hard to customise for an average user instead we could provide more pre-configured options if there are any other frequently used formats.

GimleLarpes replied:
AFAIK dayjs is already a hard dependancy, as the pre-configured options are also dayjs strings? There are so many possible preferences regarding things like seperators (ex. / - . : | space (even nothing in some languages)) that I don't think it would be feasible to provide a comparable amount of choice. 
The ability to set a custom date format was also specifically requested and seemingly well received over on [#60](https://github.com/cinnyapp/cinny/issues/60), so I don't know. I personally of course would like to have the functionality, and I don't think it's that difficult to customize since there are basically "templates" included. But I do see the concern with linking to external websites.

We could of course include more options, I just think it's cleaner to have fewer options in the dropdown.

GimleLarpes replied:
AFAIK dayjs is already a hard dependancy, as the pre-configured options are also dayjs strings? There are so many possible preferences regarding things like seperators (ex. `/` `-` `.` `:` ` ` (even nothing in some languages)) that I don't think it would be feasible to provide a comparable amount of choice.
The ability to set a custom date format was also specifically requested and seemingly well received over on #60 , so I don't know. I personally of course would like to have the functionality, and I don't think it's that difficult to customize since there are basically "templates" included. But I do see the concern with linking to external websites.

We could of course include more options as well, I just think it's cleaner to have fewer options in the dropdown.

kfiven replied:
Hard dependency meant, we can replace the pre-configured options in the code and have same functionality but not the custom ones as they will be saved in local storage and could be in any format.

In case of separator, I think `-` is the only another widely used so we can add one option for that.

GimleLarpes replied:
I see, is removing dayjs being considered? If so, then yeah, this would at least have to wait until the swap has been made. 
If not though I would like to give users extra functionality now and then in the worst case remove it later, instead of never giving it since there is a possibility it might have to be removed / updated in the future.

In the English speaking world `-` might be the only other widely used one, but the standard in my region is `.`, and quite a few seem to use `. ` and characters/kanji as standard as well, not taking individual preference into account. Adding more settings to get some customization back also feels very clunky compared to just letting people choose.

ajbura replied:
i also feel like having dayjs doc link is wrong direction. can we think of any other options about it?

GimleLarpes replied:
We could maybe have a "help" page/popup that shows the syntax without going to external websites? I know that people have been asking for a page like that for commands, so it might get reused for that as well.

ajbura replied:
maybe we can have a info menu in custom input just like we have in login screen for username field.

ajbura replied:
@GimleLarpes pinging in case you missed my feedback.^^^^^

GimleLarpes replied:
Good idea, I'll try doing something similar to the info menu.

GimleLarpes commented:
I've now added a hint for the formatting. I would've preferred to prevent overflow using `flexFlow: 'column wrap'` or something similar, but the current method works. I've also added 24hr functionality to the TimePicker.

ajbura reviewed (approved):
Looks Great! Thanks for such a nice addition.
